리사이클러뷰 + listAdapter로 영화 리스트 구현

withoutsummer opened the pull request:
리스트 어댑터로 영화리스트까지 구현하였습니다.

중간에 광고 뷰를 추가해야 하는데, 뷰 xml을 따로 추가를 하고 데이터 클래스도 따로 생성해서,
어댑터에 뷰 홀더를 추가하는게 맞을까요?

아니면 하나의 리스트 아이템 데이터 클래스에 
영화 데이터 클래스와 광고 데이터 클래스를 생성하는게 좋을지 궁금합니다!

withoutsummer committed:
리사이클러뷰 + listAdapter로 영화 리스트 구현

nathankim0 commented on android-movie/app/build.gradle.kts:
```diff
@@ -33,6 +33,10 @@ android {
     }
+
+    buildFeatures{
+        viewBinding = true
```
viewBinding 👍

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/MainActivity.kt:
```diff
@@ -5,12 +5,44 @@ import androidx.activity.enableEdgeToEdge
+            Movie(R.drawable.poster_6,"해리포터와 혼혈 왕자","2009-01-10~2009-03-01","180분"),
+            Movie(R.drawable.poster_7,"해리포터와 죽음의 성물 I","2010-01-10~2010-03-01","180분"),
+            Movie(R.drawable.poster_8,"해리포터와 죽음의 성물 II","2011-01-10~2011-03-01","180분")
+        )
```
movieList 를 onCreate 바깥에서 선언해주세요!
onCreate()는 액티비티가 생성될 때마다 호출됩니다
리스트를 onCreate() 안에 선언하면 액티비티가 재생성될 때마다(예: 화면 회전) 리스트가 새로 생성되어 불필요한 메모리 사용이 발생합니다.
또, 데이터는 UI 로직과 분리되어야 합니다. 
onCreate 바깥에 구현하시거나, viewModel 을 만들어서 분리해주시길 바랍니다.

withoutsummer replied:
아하 알겠습니다!
데이터와 UI 로직을 분리하기 위해서 리스트를  onCreate() 바깥으로 분리했습니다👍🏻

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/MainActivity.kt:
```diff
@@ -5,12 +5,44 @@ import androidx.activity.enableEdgeToEdge
+
+        //val adList : List<Ads> = listOf(
+//            Ads(R.drawable.ic_launcher_background,1)
+//        )
```
pr 을 올릴때 주석은 지워주시길 바랍니다!
좋은 코드는 주석 없이도 그 의도와 동작이 명확해야 합니다

withoutsummer replied:
주의하겠습니다! 주석은 모두 지웠습니다😊

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/MainActivity.kt:
```diff
@@ -5,12 +5,44 @@ import androidx.activity.enableEdgeToEdge
+
+        recyclerViewAdapter.submitList(movieList.toMutableList())
+        binding.mainRecyclerview.apply {
+            layoutManager = LinearLayoutManager(applicationContext)
```
```
<androidx.recyclerview.widget.RecyclerView
 app:layoutManager="androidx.recyclerview.widget.LinearLayoutManager"
 android:orientation="vertical" />
```

이렇게 하는 방법도 있습니다!

withoutsummer replied:
알려주신 방법이 기존보다 간단하게 설정이 가능해지는 것 같아요! 이 부분도 다시 수정했습니다😃

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/MainActivity.kt:
```diff
@@ -5,12 +5,44 @@ import androidx.activity.enableEdgeToEdge
+//            Ads(R.drawable.ic_launcher_background,1)
+//        )
+
+        recyclerViewAdapter.submitList(movieList.toMutableList())
```
toMutableList 는 왜 하셨을까요? 클로이의 의도가 궁금합니다!

withoutsummer replied:
아 제가 이해한걸로는 submitList가 기본적으로 불변리스트를 사용하는데, 
데이터가 변경 될 경우, diffutil을 사용하여 변경 사항을 감지하기 위해서는 다른 참조를 가져야 비교가 가능하다고 들었습니다! 
그래서 toMutableList()를 호출해서 새로운 리스트를 생성하면 데이터 변경 사항을 감지할 수 있다고 해서 코드를 구현했습니다!
근데 생각해보니 제가 작성한 데이터는 변경이 안되는 리스트이기 때문에 생략하는게 맞다고 생각이 드네요..이 부분도 수정했습니다!(사실 정확하게는 이해를 잘 못한 것 같아요..)

nathankim0 replied:
@withoutsummer movieList 가 변경되었음을 toMutableList 로 알릴 필요는 없습니다! 
diffutil을 정확하게 이해하시는 것이 필요할 것 같은데요,

1. 실제로 movieList 가 변경되었다면, 어차피 데이터가 변경되어있는 것을 diffUtil 에서 알아차릴 수 있기때문에 toMutableList가 필요가없습니다.
2. toMutableList 를 해버리면 아에 다른 객체가 넘어가기 때문에 예상치 못한 버그가 생길 수 있습니다.
3. diffUtil는 데이터의 변경된 부분만 UI 를 바꿀 수 있는 아주 편리한 도구입니다. notifyItemChanged(position) 을 쓸 필요가 없게되죠.

실제 실무에서 ```변경 사항을 감지하기 위해서는 다른 참조를 가져야..``` 하기 때문에 toMutableList 를 쓰는 일은 없습니다!

withoutsummer replied:
아하 오히려 새로운 객체 생성되서 넘어가게 되면 버그를 발생시킬 수 있고, 
diffUtil에서는 toMutableList를 안해도 데이터 변경을 알아차릴 수 있는 거였군요! 
확실하게 이해하게 된 것 같아요 감사합니다😄

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/Movie.kt:
```diff
@@ -0,0 +1,8 @@
+    val title : String,
+    val releaseTime : String,
+    val runtime: String
+)
```
model 정의 👍

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/RecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,47 @@
+
+    companion object{
+
+        val diffUtil = object : DiffUtil.ItemCallback<Movie>(){
```
diffUtil 사용 👍

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/RecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,47 @@
+            override fun areContentsTheSame(oldItem: Movie, newItem: Movie): Boolean {
+                return oldItem == newItem
+            }
+        }
```
areItemsTheSame 과 areContentsTheSame 의 차이를 공부해서 댓글로 달아주시길 바랍니다!

withoutsummer replied:
*areItemsTheSame : 기존 어댑터와 새롭게 변경되는 어댑터의 아이템이 같은지 확인한다. 주로 고유 ID를 비교해서 같은 항목인지 판단한다. 리턴 타입은 Boolean
*areContentsTheSame : 기존 어댑터와 변경되는 어댑터의 아이템 안의 내용을 비교한다. areItemsTheSame 에서 true가 나올 경우 추가적으로 비교하기 위해서 사용하는 함수

*광고 클래스와 영화 클래스를 다루는 리스트 아이템을 비교할 경우 둘의 속성이 달라서 제가 전체 필드를 확인하도록 수정했는데, 이 부분은 다시 생각해보고 수정해보겠습니다! 
<img width="684" alt="스크린샷 2024-12-31 오전 1 28 05" src="https://github.com/user-attachments/assets/a468a9ea-9b35-44ac-af17-fccf5b450c4f" />

nathankim0 replied:
@withoutsummer 보통 아이템이 같은지 확인할때는 id 만 가지고 비교하곤 합니다:) 잘 설명해주셨어요 ㅎㅎ

withoutsummer replied:
아하 그렇군요! 알겠습니다😊

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/RecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,47 @@
+            }
+        }
+    }
+}
```
줄 끝에는 개행을 넣어주시길 바랍니다!

[EOL을 넣어야 하는 이유](https://velog.io/@jangws/EOL%EC%9D%84-%EB%84%A3%EC%96%B4%EC%95%BC-%ED%95%98%EB%8A%94-%EC%9D%B4%EC%9C%A0%EC%99%80-git%EC%97%90%EC%84%9C-CRLF-EOL-%EC%B0%A8%EC%9D%B4%EB%A1%9C-%EC%9D%B8%ED%95%9C-%EB%AC%B8%EC%A0%9C-%ED%95%B4%EA%B2%B0-%EB%B0%A9%EB%B2%95)

withoutsummer replied:
수정하였습니다!🫡

nathankim0 commented on android-movie/app/src/main/res/layout/activity_main.xml:
```diff
@@ -7,13 +7,16 @@
+    </androidx.recyclerview.widget.RecyclerView>
+
 
 </androidx.constraintlayout.widget.ConstraintLayout>
```
마찬가지로 개행을 넣어주세요!

withoutsummer replied:
개행 추가로 넣었습니다👍🏻

nathankim0 commented on android-movie/app/src/main/res/values/colors.xml:
```diff
@@ -2,4 +2,6 @@
 <resources>
     <color name="black">#FF000000</color>
     <color name="white">#FFFFFFFF</color>
+    <color name="forestgreen">#2b3e34</color>
+    <color name="lichen">#CFD1B8</color>
```
color resource 활용 좋네요!

nathankim0 commented on android-movie/app/src/main/res/values/strings.xml:
```diff
@@ -1,3 +1,5 @@
 <resources>
     <string name="app_name">Movie</string>
+    <string name="movie_poster">poster</string>
+    <string name="btn_reservation">지금 예매</string>
```
string resource 활용 👍

nathankim0 commented on android-movie/app/src/main/res/layout/toolbar.xml:
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.appcompat.widget.Toolbar xmlns:android="http://schemas.android.com/apk/res/android"
```
요구사항에 ```Toolbar``` 사용X 가 적혀있습니다!
직접 구현해주시길 바랍니다

withoutsummer replied:
넵! 툴바 xml로 직접 구현해서 리팩토링하였니다😊

nathankim0 reviewed (changes requested):
열심히 구현해주셨네요 :)
몇가지 리뷰 남겼는데 확인해주시고 답변과 수정 부탁드립니다!

# 1. 질문에 대한 답변
>중간에 광고 뷰를 추가해야 하는데, 뷰 xml을 따로 추가를 하고 데이터 클래스도 따로 생성해서,
어댑터에 뷰 홀더를 추가하는게 맞을까요?
아니면 하나의 리스트 아이템 데이터 클래스에
영화 데이터 클래스와 광고 데이터 클래스를 생성하는게 좋을지 궁금합니다!

adapter 에는 하나의 모델만을 허용합니다. ```데이터 클래스도 따로 생성해서,
어댑터에 뷰 홀더를 추가하는게 맞을까요?``` 가 앞서 말씀드린 어뎁서의 속성에 따르려면 어떻게 해야할까요? 
방법은 여러가지가 있습니다

## Sealed Class 사용
```kotlin
sealed class ListItem {
    data class MovieItem(
        val title: String,
        val description: String,
        val rating: Float
        // 기타 영화 관련 데이터
    ) : ListItem()
    
    data class AdItem(
        val adTitle: String,
        val adDescription: String,
        val adImage: String
        // 광고 관련 데이터
    ) : ListItem()
}

val items = listOf(
    ListItem.MovieItem("영화 제목", "영화 설명", 4.5f),
    ListItem.AdItem("광고 제목", "광고 설명", "광고 이미지 URL"),
    ListItem.MovieItem("다른 영화", "다른 설명", 3.8f)
)
```

## viewType 을 갖고있는 경우
```kotlin
enum class ViewType {
    MOVIE,
    AD
}

data class Item(
    val viewType: ViewType,
    val movie: Movie?, // nullable
    val ad: Ad?       // nullable
)
```


Sealed class 사용과 viewtype을 갖고있는경우의 장단점을 ChatGpt 를 활용해서 정리해서 올려주시고, 클로이가 생각했을때 더 나은 방법을 가지고 Adapter 를 구현해보시길 바랍니다!

---

# 2. 컨벤션
## 코드 스타일 컨벤션
기본: [PRNDCompany 컨벤션](https://github.com/PRNDcompany/android-style-guide)

## git 컨벤션

### 브랜치 컨벤션

- 브랜치를 새로 만들기 전에, 반드시 이슈를 먼저 만들도록 한다. (교육시에는 적용X)
- 브랜치 명은 연관된 이슈 번호를 반드시 붙인다. (교육시에는 적용X)
- 브랜치는 1회용이다. 한번 머지된 브랜치를 재사용하지 말자.

### 커밋 메시지 컨벤션

- 커밋 메시지는 가능한 명령으로 작성한다.
- 커밋 헤더를 붙인다.
  - feat: 신규 기능 구현 : 사용자를 위한 신 기능; 새로운 빌드 관련 기능은 미포함
  - fix: 버그나 에러 수정 : 사용자 버그 수정; 빌드 관련 버그 픽스 미포함
  - refactor: 리팩토링 : Production Code ( 비즈니스 로직 ?)의 수정사항; 변수의 이름 변경 포함.
  - style: 코드 스타일 수정 : 포멧, 잊어버린 세미 콜론 등; Production Code의 코드 변경이 없음.
  - docs: 문서 작업 : 문서의 변경점
  - test: 테스트 코드 : 빠진 테스트의 추가, 테스트의 리팩터링; Production Code의 변경 없음
  - resource: 이미지, json 파일 등 추가, 편집 등의 변경사항
  - chore: 그외 기타 : updating grunt tasks etc; Production Code의 변경 없음

### 코드 리뷰 컨벤션

머지를 하기 전에는 반드시 approve 를 리뷰로 달아야 한다.
시간이 부족하더라도 아래 사항들은 리뷰어가 꼭 확인해야 한다.

1. 앱이 정상적으로 빌드가 되는가?
2. 앱이 정상적으로 AVD 혹은 Android device 에서 실행되는가?
3. ```./gradlew ktlintCheck``` 또는 ```./gradlew ktlintFormat``` 을 했을 때 lint 를 통과하는가?

---

# 3. PR 형식
1. 기능구현 요약
2. 오류상황
3. 질문
4. 동작 스크린샷 또는 동영상 첨부

withoutsummer committed:
refactor,feat

refactor :  sealed class 사용해 데이터 클래스 구현, movieList 선언 위치 수정, 주석 삭제, LinearLayoutManager 설정 xml로 변경,  개행 추가, 툴바 직접 구현

feat: bottomSheetDialogFragment사용하여 극장 선택 기능 구현, NestedScrollView를 사용하여 영화 상세 페이지 구현

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/BottomSheetFragment.kt:
```diff
@@ -0,0 +1,53 @@
+    ): View? {
+        binding = FragmentBottomSheetBinding.inflate(inflater,container,false)
+
+        val theaterName1 = binding.tvTheater1.text.toString()
+        val theaterName2 = binding.tvTheater2.text.toString()
+        val theaterName3 = binding.tvTheater3.text.toString()
```
theater 는 api 로 받아오는 상황이라면 많아 질 수도 있습니다! recyclerview 로 만들어주는게 좋지만, 일단 이번 범위는 아니므로 이렇게 생각할 수도 있겠다~ 만 아시면 되게겠습니다.

withoutsummer replied:
아하 알겠습니다! 다음에 기회가 된다면 수정해보겠습니다😊

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/BottomSheetFragment.kt:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    private fun navigationToCompletedPage(theaterName : String){
+        val intent = Intent(requireContext(),ReservationCompletedActivity::class.java).apply {
+            putExtra("theater_name",theaterName)
+        }
+        startActivity(intent)
+        dismiss()
```
잘 구현해주셨습니다!
추가로 생각해볼 거리를 드리자면, Intent를 만드는 역할과 책임은 이동할 Activity에 있지 않을까요? 
따라서 이렇게 만드는 방법도 있습니다.

```kotlin
class ReservationCompletedActivity : AppCompatActivity() {
    companion object {
        private const val EXTRA_THEATER_NAME = "theater_name"
        
        fun createIntent(context: Context, theaterName: String): Intent {
            return Intent(context, ReservationCompletedActivity::class.java).apply {
                putExtra(EXTRA_THEATER_NAME, theaterName)
            }
        }
    }
}
```

이렇게 하면 2가지의 장점이 있습니다

1. 캡슐화와 재사용성
- Activity를 시작하는 코드가 여러 곳에서 필요할 수 있는데, companion object에 두면 한 곳에서 관리할 수 있습니다.
- 나중에 intent에 추가 데이터가 필요하거나 로직이 변경되어도 한 곳만 수정하면 됩니다.

2. 일관성 유지
- 여러 곳에서 같은 Activity를 시작할 때 일관된 방식으로 데이터를 전달할 수 있습니다.
- 실수로 잘못된 키값("theater_name")을 사용하거나 다른 형식의 데이터를 전달하는 실수를 방지할 수 있습니다.

withoutsummer replied:
제시해주신 방법이 나중에 코드 수정시에도 간편하고 재사용하기에도 좋은 방법인 것 같아요!
해당 방법으로 기존 코드 다시 수정했습니다👍🏻

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/ListItem.kt:
```diff
@@ -0,0 +1,16 @@
+        val adImg : Int,
+        val adUrl : String
+    ):ListItem()
+}
```
sealed class 활용 👍

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/MainActivity.kt:
```diff
@@ -1,20 +1,69 @@
+        ListItem.MovieItem(6, R.drawable.poster_6,"해리포터와 혼혈 왕자","상영일 : 2009-01-10~2009-03-01","러닝타임 : 180분"),
+        ListItem.AdItem(R.drawable.toss_ad,"https://www.dreamfora.com/"),
+        ListItem.MovieItem(7, R.drawable.poster_7,"해리포터와 죽음의 성물 I","상영일 : 2010-01-10~2010-03-01","러닝타임 : 180분"),
+        ListItem.MovieItem(8, R.drawable.poster_8,"해리포터와 죽음의 성물 II","상영일 : 2011-01-10~2011-03-01","러닝타임 : 180분")
```
여기서 3가지의 문제점을 생각해볼 수 있습니다.
1. ```상영일```, ```러닝타임```: 등의 단어는 보통 api 에서 주어지지 않습니다. 
2. 이 부분을 다국어 처리하려고 하면 어떻게 될까요?
3. 디자이너가 ```상영일``` 이 부분만 다른 UI 로 바꿔달라고 하면 대응이 어렵지 않을까요?

고민해보시고 xml 과 데이터를 수정해주시길 바랍니다!

withoutsummer replied:
그렇네요!
상영 시간과 러닝타임의 레이블을 텍스트뷰로 분리하여 xml과 데이터를 다시 수정했습니다!

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/MainActivity.kt:
```diff
@@ -1,20 +1,69 @@
+                    putExtra("release_time",movieItem.releaseTime)
+                    putExtra("runtime",movieItem.runtime)
+                    putExtra("poster",movieItem.poster)
+                    putExtra("seriesNumber",movieItem.series)
```
[매직 리터럴, 매직 넘버는 전부 상수로 바꿔주시길 바랍니다.](https://javabom.tistory.com/28)

withoutsummer replied:
수정 완료했습니다🫡

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/MovieDetailActivity.kt:
```diff
@@ -0,0 +1,99 @@
+        }
+
+        binding.cvMinusBtn.setOnClickListener {
+            if(ticketsCount > 1){
```
객체지향 생활체조 원칙의 early return 을 상용해주시길 바랍니다!

withoutsummer replied:
when을 이용해서 수정했습니다😊

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/MovieDetailActivity.kt:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    private fun updateTicketsCount(){
+        binding.tvTicket.text = ticketsCount.toString()
```
역할에 맞게 매서드 분리를 잘해주셨네요!!

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/RecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,100 @@
+class RecyclerViewAdapter(
+    private val onMovieClickListener: (ListItem.MovieItem) -> Unit,
+    private val onButtonClickListener : (ListItem.MovieItem) -> Unit,
+    private val onAdClickListener : (ListItem.AdItem) -> Unit
```
Adapter 와 뷰단의 결합도가 강해보입니다! 그리고 여기서는 적지만, adapter 에서 콜백이 많게는 10몇개까지 늘어날수도 있는데 생성자가 너무 복잡해지고, 각 콜백의 용도와 책임이 모호해보입니다. 
```kotlin
interface MovieAdapterListener {
    fun onMovieClick(movie: ListItem.MovieItem)
    fun onButtonClick(movie: ListItem.MovieItem)
    fun onAdClick(ad: ListItem.AdItem)
}
```
위와같이 interface 로 분리시키는 리팩토링을 진행해주세요!

withoutsummer replied:
리팩토링 진행했습니다!👍🏻

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/RecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    companion object{
+        private val MOVIE_ITEM_TYPE = 0
+        private val AD_ITEM_TYPE = 1
+        private val TYPE_ERROR = "타입 에러"
```
companion object 에 선언하는 것도 괜찮지만, viewtype 이라는 분명한 역할이 있으므로 enum class 로 분리시키는것이 좋아 보입니다!

withoutsummer replied:
네! enum class로 분리시켰습니다!👍🏻

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/RecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,100 @@
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
+
+        return when(viewType){
+            MOVIE_ITEM_TYPE -> {
+                val view = ItemMovieBinding.inflate(LayoutInflater.from(parent.context),parent,false)
+                MovieItemHolder(view)
+            }
+            AD_ITEM_TYPE -> {
+                val view = ItemAdBinding.inflate(LayoutInflater.from(parent.context),parent,false)
+                AdItemHolder(view)
+            }
+            else -> throw IllegalArgumentException(TYPE_ERROR)
+        }
```
viewType 에 따라 잘 나누어주셨네요 :)

nathankim0 commented on android-movie/app/src/main/res/layout/back_toolbar.xml:
```diff
@@ -0,0 +1,30 @@
+        app:layout_constraintTop_toTopOf="parent" />
+
+
+</androidx.constraintlayout.widget.ConstraintLayout>
```
eol 에러가 아직 보이네요 ㅠㅠ

withoutsummer replied:
수정했습니다!!

nathankim0 reviewed (changes requested):
안녕하세요 클로이! 
리뷰드린 부분들 잘 구현해주셨네요:) 기기에서도 잘 동작하는 것 확인했습니다!! 정말 고생 많으셨습니다:)☺️

몇가지 피드백을 좀 드리자면,

#### 1. 아직 제가 질문드린 부분에 대한 답변을 주시지 않으셨습니다! 리뷰어가 리뷰한 내용은 모두 꼼꼼히 봐주시고, 또 리뷰 conversation마다 이모지나 짧은 답변등을 남겨주시면 더 좋은 개발 문화를 만들 수 있을 것 같습니다:)🔥
https://insight.infograb.net/blog/2023/07/20/growing-codereview/
> Sealed class 사용과 viewtype을 갖고있는경우의 장단점을 ChatGpt 를 활용해서 정리해서 올려주시고...

#### 2. commit 을 한 commit 에 몰아서 해버리면 중간에 잘못된 부분으로 돌아가기도 힘들뿐더러 커밋단위로 리뷰하기가 어려워집니다. commit 은 적절하게 나눠서 작업 중간중간 해주시면 좋겠습니다:) 🏋️‍♀️
> <img width="825" alt="image" src="https://github.com/user-attachments/assets/4cebc2e4-5e3e-4b20-9834-57176d2b525c" />



정말 잘 구현해주셨지만 추가로 몇가지 더 리뷰드렸는데 반영을 해주시고 다시 요청 부탁드려요~🚀

nathankim0 commented:
@ghojeong 클로이 코드 자랑해봅니다! 폭풍 성장하고 있답니다ㅋㅋ 혹시 리뷰하실 내용있다면 남겨주세요ㅎㅎ

ghojeong commented on android-movie/app/src/main/java/com/example/movie/MovieDetailActivity.kt:
```diff
@@ -0,0 +1,99 @@
+import androidx.core.view.WindowInsetsCompat
+import com.example.movie.databinding.ActivityMovieDetailBinding
+
+class MovieDetailActivity : AppCompatActivity() {
+
+    private lateinit var binding: ActivityMovieDetailBinding
+
+    val movieDetailList = listOf(
+        "친척집에서 구박받는 생활을 하던 해리는 11살 생일을 앞두고 호그와트 마법학교로부터 입학초대장을 받고 자신이 마법사라는 " +
+                "사실을 알게 된다. 해리는 호그와트 마법학교로 가는 열차에서 친구 론과 헤르미온느를 사귀고 함께 마법과 신비, " +
+                "모험으로 가득한 학교생활을 시작한다.",
```
movieDetailList 가 Activity 안에 하드코딩되어 박혀있는데,
하드코딩된 데이터라 할지라도 별도 객체로 분리를 해주시면 좋습니다.
하드코딩된 데이터는 언젠가는 xml 파일을 통해서, 혹은 로컬 DB 를 통해서, 혹은 서버를 통해서 가져와야할 대상이니까요

withoutsummer replied:
아하 알겠습니다!
별로 객체로 분리하였습니다🫡

ghojeong commented on android-movie/app/src/main/java/com/example/movie/ReservationCompletedActivity.kt:
```diff
@@ -0,0 +1,34 @@
+
+
+
+        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
+            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
+            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
```
그 어떤 경우더라도 알파벳 하나짜리 파라미터 혹은 변수명을 사용하지 말아주세요.
코틀린에서 익명 파라미터인 it 를 허용하는 이유는 문맥과 it 의 타입추론을 통해서,
it 가 무엇인지 코드를 읽는 동료 또한 추측을 하기 쉽기 때문입니다.

만약 it 를 사용할 수 없는 상황이라면, 코틀린 컨벤션상 제대로 파라미터 이름을 지어주지 않으면,
가독성에 문제가 있는 상황이라고 컴파일러가 경고를 주는 거라고 생각해도 좋습니다.

v 파라미터를 제대로 이름지어주세요.

withoutsummer replied:
아 이 부분은 프래그먼트를 생성했을 때, 기존에 구현되어 있었던 코드여서 확인을 못했었는데 수정해보겠습니다!

ghojeong reviewed (dismissed):
안녕하세요 클로이!
정말 열심히 과제를 하고 계시군요! 놀러왔습니다.
읽으면서 사소한 의견 몇개 남겨보았습니다.

withoutsummer committed:
refactor : intent 역할과 책임 이동

refactor : ReservationCompletedActivity와 MovieDetailActivity에 Intent 생성을 담당하는 메서드를 각각 추가하여 역할과 책임 이동

withoutsummer committed:
refactor : 상영시간, 러닝타임 데이터 레이블 분리 및 데이터 수정

refactor : list_movie.xml과 activity_movie_detail.xml의 상영시간과 러닝타임 레이블을 텍스트뷰로 분리하고 데이터를 수정하였습니다!

withoutsummer committed:
refactor : MovieDetailActivity.kt 하드코딩된 movieDetailList 별로 객체로 분리, 뷰타입 enum class로 분리, eol 에러 수정

withoutsummer committed:
1. refactor : RecyclerViewAdapter.kt 에서 어댑터 콜백 함수를 interface로 분리, ReservationCompletedActivity.kt 에서 setOnApplyWindowInsetsListener의 v 파라미터 변수명 수정       2.feat : Activity_reservation_completed.xml 레이아웃 추가, ReservationCompletedActivity.kt에서 플래그먼트에서 영화 제목, 극장 이름 데이터를 intent로 전달받아 뷰바인딩 기능 추가

withoutsummer committed:
refactor : onCreateViewHolder()에서 else없이 뷰타입 구분할 수 있도록 수정

withoutsummer commented:
## sealed class(봉인 클래스)
 특정 타입의 계층 구조를 표현하기 위한 클래스
이를 통해 컴파일러가 타입을 완전히 알고 있어야 하는 상황에서 강력한 타입 안정성을 제공

### 특징
1.  계층 구조 표현 : 상속이 가능한 클래스이지만, 모든 하위 클래스는 같은 파일 내에 정의되어야 한다. 
    -> 이를 통해 특정 타입 계층을 안전하게 표현할 수 있다.
2. 타입 안정성 : when구문에서 모든 하위 클래스를 다루지 않으면 컴파일 오류 발생한다.
   -> 이를 통해 누락된 타입 처리를 방지할 수 있음
3. 데이터 포함 기능 : 각 하위 클래스는 데이터를 포함하거나 고유한 메서드를 가질 수 있다.
4. 확장성 : 새로운 하위 클래스를 추가해도 기존 코드를 수정할 필요가 없다. -> 유지보수성이 좋다!

### 장점
1. 타입 안정성 : 컴파일러가 하위 타입을 모두 알고 있어, 처리되지 않은 타임을 방지할 수 있다.
2. 유지보수 용이 : 새로운 하위 타입을 추가해도 기존 코드가 영향 받지 않는다.
3. 데이터와 동작을 함께 관리 : 하위 클래스마다 다른 데이터를 포함하거나 동작을 정의할 수 있다.
4. 복잡한 구조를 표현 가능 : 타입마다 고유한 동작과 데이터를 다룰 수 있어 유연하다.

### 단점
1. 파일 내부에 제한 : 모든 하위 타입은 같은 파일에 정의되어야 하므로 대규모 프로젝트에서 구조적 번거로움이 있을 수 있음!
2. 타입 비교의 복잡성 : 타입 구분을 위해 `when` 또는` is` 를 사용해야 하므로, 간단한 데이터 구조에는 적합하지 않다!

-------
## enum class(열거형 클래스)
 미리 정의된 값들의 고정된 집합을 표현하기 위해 사용되는 클래스
이는 각 값이 상수로 정의 되며, 보통 타임 구분과 값 비교를 위한 용도로 사용된다.

### 특징
1. 값의 집합 정의 : 고정된 값의 집합을 정의한다.
5. 상수 값 관리 : enum의 각 값은 상수로 정의된다. 즉, 상수 값들은 모두 하나의 enum 타입으로 취급!
6. 고유 메서드 정의 기능 : 각 상수 값에 고유한 속성이나 메서드를 추가할 수 있다.
7. 빠른 비교 : enum 값을 상수로 저장하므로, 간단히 == 로 비교 가능!
8. 타입 안전성 : 특정 값만 허용되는 상황에서 타입 안정성을 제공한다.

### 장점
1. 간단한 타입 구분 : enum 값만으로 타입을 구분할 수 있다.
2. 빠른 비교 : `== `연산으로 상수 값을 비교하기 수월!
4. 간결한 구조 : 데이터를 포함하거나 복잡한 계층 구조를 정의하지 않아도 된다.
5. 일관된 값 관리 : 정의된 값 외의 데이터가 들어오는 것을 방지한다.

### 단점
1. 확장성 부족 : 새로운 타입을 추가하면 기존 enum 정의와 모든 관련 코드를 수정해야 함.
2. 데이터 구조 표현의 한계 : 타입별로 고유 데이터를 포함하거나 동작을 정의할 수 없다.
3. 고정된 값 집합 : 런타임에 새로운 값을 추가할 수 없다.

-----
 ### 선택 기준

- sealed class를 사용하는 경우 : 타입마다 다른 데이터와 동작을 정의해야 하는 복잡한 데이터 구조
- enum class를 사용하는 경우 : 단순히 타입이나 상태를 구분해야 하는 경우

-
@nathankim0  Sealed class 사용과 viewtype을 갖고있는경우의 장단점을 ChatGpt 를 활용해서 정리했습니다!

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/ViewType.kt:
```diff
@@ -0,0 +1,13 @@
+
+    companion object{
+        fun fromInt(type : Int):ViewType {
+            return values().find { it.type == type }
```
values() 대신 ```entries``` 를 쓰기도 합니다~

```'Enum. values()' is recommended to be replaced by 'Enum. entries' since 1.9 ```

withoutsummer replied:
아하 찾아보니, values는 가변인 배열로 반환돼서 중간에 수정이 되거나 순서가 바뀔 수 있어 일관성 유지가 어려운 반면에, entries를 사용하면 불변인 list로 반환하여 더 안전할 것 같네요! 
이 부분 참고해서 수정하겠습니다☺️

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/RecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,95 @@
+
+    override fun getItemViewType(position: Int): Int {
+        return when (getItem(position)) {
+            is ListItem.MovieItem -> ViewType.MOVIE_ITEM.type
+            is ListItem.AdItem -> ViewType.AD_ITEM.type
```
enum 활용해서 잘 바꾸어주셨네요 :) !!!

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/RecyclerViewAdapter.kt:
```diff
@@ -0,0 +1,95 @@
+import com.example.movie.databinding.ItemMovieBinding
+
+class RecyclerViewAdapter(
+    private val listener: MovieAdapterListener
```
listener 로 리팩토링 잘해주셨네요!!!!

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/MovieDetailActivity.kt:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    companion object {
+        private const val EXTRA_TITLE = "title"
+        private const val EXTRA_RELEASE_TIME = "release_time"
+        private const val EXTRA_RUNTIME = "runtime"
+        private const val EXTRA_POSTER = "poster"
+        private const val EXTRA_SERIES_NUMBER = "seriesNumber"
```
상수화 좋습니다:)

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/MovieDetailActivity.kt:
```diff
@@ -0,0 +1,100 @@
+        }
+
+        binding.cvMinusBtn.setOnClickListener {
+            when {
+                ticketsCount > 1 -> {
+                    ticketsCount--
+                    updateTicketsCount()
+                }
```
if 대신 early return 를 사용하는 것이 좋다고 리뷰 드렸었는데
if 와 when 은 같은의미나 마찬가지입니다! 
```kotlin
if(ticketsCount <= 1) return@setOnClickListener
ticketsCount--
updateTicketsCount()
```
예시 보여드립니다:) 다음 단계에서 기억해주세요~~

withoutsummer replied:
아하 기존과 비교해보니 early return을 사용하는 방법이 if 조건이 참일 경우에 바로 리턴되어서 불필요한 코드 실행이 안되어 성능 향상에도 더 좋은 방법인 것 같습니다!

nathankim0 replied:
@withoutsummer 첨언하자면 성능 향상과는 크게 관련이 없습니다 ㅎㅎ

nathankim0 replied:
@withoutsummer 그냥 조건문에 들어가지 않게 해서 읽기 쉽게 만들어 줄 뿐입니다
근데 early return 은 사람마다 의견이 좀 달라서 링크 첨부해드리니 한번 읽어보시면 좋을 것 같아요

-> https://thearchivelog.dev/article/are-early-returns-any-good/

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/MovieDetailActivity.kt:
```diff
@@ -0,0 +1,100 @@
+        binding = ActivityMovieDetailBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+
+        val seriesNumber = intent.getIntExtra(EXTRA_SERIES_NUMBER, -1)
+        val posterResId = intent.getIntExtra(EXTRA_POSTER, -1)
+        val title = intent.getStringExtra(EXTRA_TITLE) ?: "Unknown Title"
+        val releaseTime = intent.getStringExtra(EXTRA_RELEASE_TIME) ?: "Unknown Release Time"
+        val runtime = intent.getStringExtra(EXTRA_RUNTIME) ?: "Unknown Runtime"
```
크으 intent 로 예상치 못한 값이 들어왔을때 처리를 잘해주셨네요:) 
보통 throw 해서 앱을 팅겨버리게 하는 분들도 있는데요!
자랑스럽네요~~

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/MainActivity.kt:
```diff
@@ -1,20 +1,118 @@
+    private lateinit var recyclerViewAdapter: RecyclerViewAdapter
+
+    val movieList = listOf(
+        ListItem.MovieItem(
+            1,
+            R.drawable.poster_1,
+            "해리포터와 마법의 돌",
+            "2001-01-10~2001-03-01",
+            "150분"
+        ),
```
문자 분리 잘해주셨네요!!

nathankim0 commented on android-movie/app/src/main/java/com/example/movie/MovieDetailList.kt:
```diff
@@ -0,0 +1,48 @@
+package com.example.movie
+
+object MovieDetailList {
```
object 로 분리해주셨네요~~ 나중에는 viewmodel 에서 list 를 가져오게 될 것 입니다! 기대해주세요 ㅎㅎ

nathankim0 reviewed (approved):
잘 돌아가는것과 리뷰 반영하신 부분 확인했습니다!!
정말 고생 많으셨어요 클로이ㅎㅎ!!
정말 빠른 속도로 성장하는 모습이 인상깊네요 👏
요구사항을 잘 반영하신 부분도 상당히 멋집니다.
구두로 말씀드린, 커밋메시지 작성과 협업에 대한 피드백도 잘 지켜주시리라 믿어 의심치 않습니다!
다음 단계에서도 화이팅입니다!!
